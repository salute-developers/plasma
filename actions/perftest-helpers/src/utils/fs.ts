import * as fs from 'fs';

export async function readJson<T>(path: string): Promise<T> {
    return JSON.parse(await fs.promises.readFile(path, 'utf-8'));
}
