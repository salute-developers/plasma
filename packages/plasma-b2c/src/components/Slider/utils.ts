import { MutableRefObject } from 'react';

/**
 * Расчитать значение слайдера с учетом его координат и шага изменений.
 * @param {number} handleCenterXRelative
 * @param {number} stepSize
 * @param {number} min
 * @param {number} max
 * @return {number}
 */
export function getSliderThumbValue(handleCenterXRelative: number, stepSize: number, min: number, max: number) {
    const newValue = Math.round(handleCenterXRelative / stepSize) + min;

    return Math.min(Math.max(newValue, min), max);
}

/**
 * Расчитывает значение отступа слайдера на основе его положения (справа, слева) на отрезке слайдера.
 * Значение используется для правильного расчета ограничения движения слайдера.
 * @param ref
 * @param {'left' | 'right'} side
 * @return Array<number>
 */
export function getOffsets(
    ref: ((instance: HTMLDivElement | null) => void) | MutableRefObject<HTMLDivElement | null> | null,
    side?: 'left' | 'right',
): number[] {
    if (!ref || !('current' in ref) || !ref.current || !side) {
        return [0, 0];
    }

    const size = ref.current.clientWidth;

    if (side === 'left') {
        return [0, size];
    }

    if (side === 'right') {
        return [size, 0];
    }

    return [0, 0];
}
