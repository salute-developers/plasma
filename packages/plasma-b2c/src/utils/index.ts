export * from '@salutejs/plasma-web/utils';
