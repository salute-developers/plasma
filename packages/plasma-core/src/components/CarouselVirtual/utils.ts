import { useCallback, useState } from 'react';

import { ScrollAxis, ScrollAlign } from './types';
import { useRemoteListener } from './useRemoteListener';

/**
 * Подсчет смещения из-за паддингов.
 */
export const getCalculatedOffset = (scrollEl: Element, trackEl: Element, axis: ScrollAxis) => {
    const paddingProp = axis === 'x' ? 'paddingLeft' : 'paddingTop';
    return parseInt(getComputedStyle(scrollEl)[paddingProp], 10) + parseInt(getComputedStyle(trackEl)[paddingProp], 10);
};

const round = (n: number) => Math.round(n * 100) / 100;

/**
 * Получить позицию (слот) айтема в каруселе.
 * Каждый айтем имеет свой слот относительно вьюпорта карусели.
 */
export const getItemSlot = (
    itemIndex: number,
    itemEnd: number,
    itemSize: number,
    scrollStart: number,
    scrollSize: number,
    scrollAlign: ScrollAlign,
    prevIndex = 0,
    offset = 0,
) => {
    /**
     * Граница и центр скролла (видимой части).
     * Смещение + размер.
     */
    const scrollEnd = scrollStart + scrollSize;
    const scrollCenter = scrollStart + scrollSize / 2;
    const itemCenter = itemEnd - itemSize / 2;

    if (scrollAlign === 'center') {
        return round((itemCenter - scrollCenter) / itemSize);
    }
    if (scrollAlign === 'start') {
        return round((itemEnd - itemSize - scrollStart) / itemSize);
    }
    if (scrollAlign === 'end') {
        return round((itemEnd - (scrollSize + scrollStart)) / itemSize);
    }
    if (scrollAlign === 'activeDirection') {
        const prevStart = offset + itemSize * prevIndex;
        const prevEnd = prevStart + itemSize;
        const prevVisible = prevEnd > scrollStart && prevStart < scrollEnd;

        if (!prevVisible) {
            if (prevIndex < itemIndex) {
                return round((itemEnd - (scrollSize + scrollStart)) / itemSize);
            }
            return round((itemEnd - itemSize - scrollStart) / itemSize);
        }
    }
    return null;
};

export function debounceByFrames<FN extends (...args: any[]) => void>(fn: FN, framesToDebounce = 0) {
    if (framesToDebounce === 0) {
        return fn;
    }

    let timeoutId: number | null = null;
    let framesCounter = 0;

    return (...args: Parameters<FN>) => {
        const tick = () => {
            if (framesCounter === framesToDebounce - 1) {
                timeoutId = null;
                framesCounter = 0;
                fn(...args);
            } else {
                framesCounter++;
                timeoutId = requestAnimationFrame(tick);
            }
        };

        if (timeoutId !== null) {
            framesCounter = 0;
            cancelAnimationFrame(timeoutId);
        }

        timeoutId = requestAnimationFrame(tick);
    };
}

// eslint-disable-next-line space-before-function-paren
export function throttleByFrames<FN extends (...args: any[]) => void>(fn: FN, framesToThrottle = 0) {
    if (framesToThrottle === 0) {
        return fn;
    }

    let isWaited = false;
    let framesCounter = 0;

    const tick = () => {
        if (framesCounter === framesToThrottle - 1) {
            isWaited = false;
            framesCounter = 0;
        } else {
            framesCounter++;
            requestAnimationFrame(tick);
        }
    };

    return (...args: Parameters<FN>) => {
        if (isWaited) {
            return;
        }

        fn(...args);
        isWaited = true;

        tick();
    };
}

const throttlingParamsDefault = {
    leading: true,
    trailing: false,
};

/**
 * Хук для навигации. Слушает нажатие кнопок на пульте/клавиатуре.
 */
export function useRemoteHandlers({
    initialIndex = 0,
    axis,
    delayFrames,
    longDelayFrames,
    min,
    max,
    count = 1,
    longCount = 5,
}: {
    initialIndex: number;
    axis: ScrollAxis;
    delayFrames: number;
    longDelayFrames: number;
    min: number;
    max: number;
    count?: number;
    longCount?: number;
    throttlingParams?: typeof throttlingParamsDefault;
}) {
    const indexState = useState(initialIndex);
    const [, setIndex] = indexState;

    const step = useCallback(
        throttleByFrames(
            (cmd: '+' | '-') =>
                setIndex((prevIndex) => {
                    if (cmd === '+') {
                        return prevIndex + count <= max ? prevIndex + count : min;
                    }
                    return prevIndex - count >= min ? prevIndex - count : max;
                }),
            delayFrames,
        ),
        [min, max],
    );
    const jump = useCallback(
        throttleByFrames(
            (cmd: '+' | '-') =>
                setIndex((prevIndex) => {
                    if (cmd === '+') {
                        return prevIndex + longCount <= max ? prevIndex + longCount : min;
                    }
                    return prevIndex - longCount >= min ? prevIndex - longCount : max;
                }),
            longDelayFrames,
        ),
        [min, max],
    );

    useRemoteListener((key, ev) => {
        ev.preventDefault();
        if (axis === 'x') {
            switch (key) {
                case 'LEFT':
                    step('-');
                    break;
                case 'RIGHT':
                    step('+');
                    break;
                case 'LONG_LEFT':
                    jump('-');
                    break;
                case 'LONG_RIGHT':
                    jump('+');
                    break;
                default:
                    break;
            }
        } else {
            switch (key) {
                case 'UP':
                    step('-');
                    break;
                case 'DOWN':
                    step('+');
                    break;
                case 'LONG_UP':
                    jump('-');
                    break;
                case 'LONG_DOWN':
                    jump('+');
                    break;
                default:
                    break;
            }
        }
    });

    return indexState;
}
