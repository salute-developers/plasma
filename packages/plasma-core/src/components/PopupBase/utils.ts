import { css } from "styled-components";
import { BasicPopupBasePlacement, PopupBasePlacement } from "./types";

export const handlePosition = (placement?: PopupBasePlacement, offset?: [number | string, number | string]) => {
    let x = '0rem';
    let y = '0rem';
    if (offset) {
        const [_x, _y] = offset;
        x = typeof _x === 'number' ? `${_x}rem` : _x;
        y = typeof _y === 'number' ? `${_y}rem` : _y;
    }

    if (!placement || placement === 'center') {
        return css`
            left: calc(50% + ${x});
            top: calc(50% - ${y});
            transform: translate(-50%, -50%);
        `;
    }

    let left;
    let right;
    let top;
    let bottom;
    const placements = placement.split('-') as BasicPopupBasePlacement[];

    placements.forEach((placement: BasicPopupBasePlacement) => {
        switch (placement) {
            case 'left':
                left = x;
                break;
            case 'right':
                right = x;
                break;
            case 'top':
                top = y;
                break;
            case 'bottom':
                bottom = y;
                break;
            default:
                break;
        }
    });

    const isCenteredX = left === undefined && right === undefined;
    const isCenteredY = top === undefined && bottom === undefined;

    return css`
        left: ${left};
        right: ${right};
        top: ${top};
        bottom: ${bottom};
        ${isCenteredX &&
        css`
            left: calc(50% + ${x});
            transform: translateX(-50%);
        `}
        ${isCenteredY &&
        css`
            top: calc(50% - ${y});
            transform: translateY(-50%);
        `}
    `;
};