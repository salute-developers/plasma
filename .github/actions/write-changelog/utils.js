import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';
import { promises as fs } from 'fs';

function getFilePath(pkg, fileName) {
    const __filename = fileURLToPath(import.meta.url);
    const __dirname = dirname(__filename);

    return resolve(__dirname, '../../..', `packages/${pkg}/${fileName}`);
}

function getDate(date) {
    const formatter = new Intl.DateTimeFormat('ru', {
        day: 'numeric',
        month: 'long',
        year: 'numeric',
    });
    const data = formatter.format(date).replace(' г.', '');

    return `(${data})`;
}

async function getPackageVersion(pkg) {
    const filePath = getFilePath(pkg, 'package.json');

    try {
        const packageJson = await fs.readFile(filePath, 'utf8');
        const { version } = JSON.parse(packageJson);

        return `# ${version}`;
    } catch (error) {
        console.error('Error reading package.json:', error);
    }
}

export async function writeChangelog(markdown, pkg) {
    try {
        if (!pkg) {
            console.error('Не указан пакет для записи об изменениях');

            return;
        }

        const version = await getPackageVersion(pkg);
        const mdFilePAth = getFilePath(pkg, 'CHANGELOG.md');
        const CHANGELOG = await fs.readFile(mdFilePAth, 'utf8');

        const changelog = `${version} ${getDate(new Date())}\n\n${markdown}\n\n${CHANGELOG}`;

        await fs.writeFile(mdFilePAth, changelog, 'utf8');

        console.log(`${pkg} - changelog успешно записан`);
    } catch (error) {
        console.error('Ошибка при записи файла:', error);
    }
}
