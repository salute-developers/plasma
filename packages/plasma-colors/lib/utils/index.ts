import fs from 'fs';
import path from 'path';

// TODO: https://github.com/salute-developers/plasma/issues/1181

export const ROBO_COMMENT = '// Generated by robots, do not change this manually!\n';

export const generateFile = (name: string, content: string | object, deprecated = '') => {
    if (typeof content !== 'string') {
        content = JSON.stringify(content, null, 4);
    }

    return {
        file: `${name}.ts`,
        content: `export const ${name} = ${content};${deprecated}\n`,
    };
};

export type GeneratedFiles = Array<{
    file: string;
    content: string;
}>;

export const writeGeneratedToFS = (dir: string, generated: GeneratedFiles, generateMessage = ROBO_COMMENT) => {
    fs.existsSync(dir) || fs.mkdirSync(dir);

    for (const { file, content } of generated) {
        const src = path.join(dir, file);

        if (fs.existsSync(src)) {
            fs.appendFileSync(src, content);
            return;
        }

        fs.writeFileSync(src, `${generateMessage}\n${content}`);
    }
};

export const getDirectoryContent = (dir: string) =>
    fs
        .readdirSync(dir, { withFileTypes: true })
        .filter((item) => item.isFile() && path.extname(path.join(dir, item.name)) === '.json')
        .reduce((colorSchemas, item) => {
            const fileContent = fs.readFileSync(path.join(dir, item.name), 'utf-8');
            const name = path.parse(item.name).name;

            return {
                ...colorSchemas,
                [name]: fileContent,
            };
        }, {});
