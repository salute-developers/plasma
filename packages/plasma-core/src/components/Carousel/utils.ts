import { animatedScrollToX, animatedScrollToY, TimingFunction } from '../../utils';

import { ScrollAxis, ScrollAlign, UseCarouselLiteOptions } from './types';

const positionModByScrollAlign = ({
    scrollAlign,
    position,
    carouselSize,
    itemSize,
    offset,
    scrollStart,
    axis,
}: {
    scrollAlign: ScrollAlign;
    position: number;
    carouselSize: number;
    itemSize: number;
    offset: number;
    scrollStart: number;
    axis: ScrollAxis;
}) => {
    if (scrollAlign === 'start') {
        const inaccuracy = 1;
        const paddingOffset = axis === 'y' ? offset - itemSize / 2 + inaccuracy : 0;
        return position + paddingOffset;
    }
    if (scrollAlign === 'center') {
        return position - carouselSize / 2 + itemSize / 2;
    }
    if (scrollAlign === 'end') {
        return position - carouselSize + itemSize + offset;
    }
    if (scrollAlign === 'activeDirection') {
        if (position >= scrollStart + carouselSize - itemSize) {
            return position - carouselSize + itemSize + offset;
        }
        if (position > scrollStart) {
            return scrollStart;
        }
    }
    return position;
};

/**
 * Подсчет скролла до переданного индекса.
 */
export const getCalculatedPos = ({
    scrollEl,
    items,
    axis,
    index,
    offset,
    scrollAlign,
}: {
    scrollEl: HTMLElement;
    items: HTMLCollectionOf<HTMLElement>;
    axis: ScrollAxis;
    index: number;
    offset: number;
    scrollAlign: ScrollAlign;
}) => {
    let position = scrollAlign === 'center' ? offset : 0;
    let carouselSize;
    let itemSize;
    let scrollStart;

    if (items.item(index) === null) {
        return position;
    }

    for (let i = 0; i < index; i++) {
        if (axis === 'x') {
            position += items.item(i)?.offsetWidth ?? 0;
        } else {
            position += items.item(i)?.offsetHeight ?? 0;
        }
    }

    if (axis === 'x') {
        carouselSize = scrollEl.offsetWidth;
        itemSize = items.item(index)?.offsetWidth ?? 0;
        scrollStart = scrollEl.scrollLeft;
    } else {
        carouselSize = scrollEl.offsetHeight;
        itemSize = items.item(index)?.offsetHeight ?? 0;
        scrollStart = scrollEl.scrollTop;
    }

    return positionModByScrollAlign({
        scrollAlign,
        position,
        carouselSize,
        itemSize,
        offset,
        scrollStart,
        axis,
    });
};

/**
 * Подсчет смещения из-за паддингов.
 */
export const getCalculatedOffset = (scrollEl: Element, trackEl: Element, axis: ScrollAxis) => {
    const paddingProp = axis === 'x' ? 'paddingLeft' : 'paddingTop';
    return parseInt(getComputedStyle(scrollEl)[paddingProp], 10) + parseInt(getComputedStyle(trackEl)[paddingProp], 10);
};

/**
 * Прокрутка к указанной позиции с анимацией или без.
 */
export const scrollToPos = ({
    scrollEl,
    pos,
    axis,
    animated,
    duration,
    timingFunction,
}: {
    scrollEl: HTMLElement;
    pos: number;
    axis: ScrollAxis;
    animated?: boolean;
    duration?: number;
    timingFunction?: TimingFunction;
}) => {
    if (axis === 'x' && Math.abs(pos - scrollEl.scrollLeft) > 1) {
        if (animated) {
            animatedScrollToX(scrollEl, pos, duration, timingFunction);
        } else {
            scrollEl.scrollTo({ left: pos });
        }
    }
    if (axis === 'y' && Math.abs(pos - scrollEl.scrollTop) > 1) {
        if (animated) {
            animatedScrollToY(scrollEl, pos, duration, timingFunction);
        } else {
            scrollEl.scrollTo({ top: pos });
        }
    }
};

const round = (n: number) => Math.round(n * 100) / 100;

/**
 * Получить позицию (слот) айтема в каруселе.
 * Каждый айтем имеет свой слот относительно вьюпорта карусели.
 */
export const getItemSlot = (
    itemIndex: number,
    itemEnd: number,
    itemSize: number,
    scrollStart: number,
    scrollSize: number,
    scrollAlign: ScrollAlign,
    prevIndex = 0,
    offset = 0,
) => {
    /**
     * Граница и центр скролла (видимой части).
     * Смещение + размер.
     */
    const scrollEnd = scrollStart + scrollSize;
    const scrollCenter = scrollStart + scrollSize / 2;
    const itemCenter = itemEnd - itemSize / 2;

    if (scrollAlign === 'center') {
        return round((itemCenter - scrollCenter) / itemSize);
    }
    if (scrollAlign === 'start') {
        return round((itemEnd - itemSize - scrollStart) / itemSize);
    }
    if (scrollAlign === 'end') {
        return round((itemEnd - (scrollSize + scrollStart)) / itemSize);
    }
    if (scrollAlign === 'activeDirection') {
        const prevStart = offset + itemSize * prevIndex;
        const prevEnd = prevStart + itemSize;
        const prevVisible = prevEnd > scrollStart && prevStart < scrollEnd;

        if (!prevVisible) {
            if (prevIndex < itemIndex) {
                return round((itemEnd - (scrollSize + scrollStart)) / itemSize);
            }
            return round((itemEnd - itemSize - scrollStart) / itemSize);
        }
    }
    return null;
};

export function getCarouselItems(track: HTMLElement): HTMLCollectionOf<HTMLElement> {
    return track.children as HTMLCollectionOf<HTMLElement>;
}

const axisToTranslateMap: Record<UseCarouselLiteOptions['axis'], (position: number) => string> = {
    x(position) {
        return `translateX(${-position}px)`;
    },
    y(position) {
        return `translateY(${-position}px)`;
    },
};
const axisToOffsetKeyMap: Record<UseCarouselLiteOptions['axis'], 'offsetLeft' | 'offsetTop'> = {
    x: 'offsetLeft',
    y: 'offsetTop',
};

const axisToSizeKeyMap: Record<UseCarouselLiteOptions['axis'], 'offsetWidth' | 'offsetHeight'> = {
    x: 'offsetWidth',
    y: 'offsetHeight',
};

const axisToScrollKeyMap: Record<UseCarouselLiteOptions['axis'], 'scrollLeft' | 'scrollTop'> = {
    x: 'scrollLeft',
    y: 'scrollTop',
};

function getCenterPosition(itemSize: number, itemStart: number, carouselSize: number, trackOffset: number) {
    const relativeMiddle = itemStart + trackOffset + itemSize / 2;

    return relativeMiddle - carouselSize / 2;
}

function getEndPosition(itemSize: number, itemStart: number, carouselSize: number, trackOffset: number) {
    return 2 * trackOffset + itemStart + itemSize - carouselSize;
}

function getTranslatePosition(
    itemSize: number,
    itemStart: number,
    carouselSize: number,
    trackOffset: number,
    scrollAlign: UseCarouselLiteOptions['scrollAlign'],
): number {
    switch (scrollAlign) {
        case 'start': {
            return itemStart;
        }
        case 'end': {
            return getEndPosition(itemSize, itemStart, carouselSize, trackOffset);
        }
        case 'center':
        case 'activeDirection': // TODO: activeDirection сделать позже, мало кто использует. Fallback на 'center'
        default:
            return getCenterPosition(itemSize, itemStart, carouselSize, trackOffset);
    }
}

function boundPosition(position: number, trackSize: number, carouselSize: number, trackOffset: number): number {
    if (position < 0) {
        return 0;
    }

    // если все элементы помещаются в ширину карусели, то считаем trackEnd по-другому
    if (trackSize < carouselSize) {
        const trackEnd = Math.abs(carouselSize - trackSize - trackOffset);

        return position < trackEnd ? position : trackEnd;
    }

    const trackEnd = trackSize - carouselSize + trackOffset;

    if (position > trackEnd) {
        return trackEnd;
    }

    return position;
}

/**
 * Функция решает являются ли переданные индексы крайними с начала и конца.
 * Например в такой карусели `([1][2][3][4])` — при изменении индекса с 4 на 1 или наоборот функция вернёт `true`,
 * так как это переход между крайними элементами.
 *
 * Есть исключение, для карусели длинной в 2 элемента `([1][2])`.
 * В этом случае переход с индекса 1 на 2 и наоборот не должен считаться переходом с одного края на другой.
 *
 * @param prevIndex
 * @param index
 * @param numberOfItems количество элементов в карусели
 */
function isMaximumDistance(prevIndex: number, index: number, numberOfItems: number) {
    // проверка, что карусель состоит из 2-х элементов
    if (numberOfItems === 2) {
        return false;
    }

    return Math.abs(index - prevIndex) === numberOfItems - 1;
}

function setTimingFunction(element: HTMLElement | null, timingFunction: 'step-start' | '') {
    if (element) {
        element.style.transitionTimingFunction = timingFunction;
    }
}

function translateToPosition(element: HTMLElement, axis: UseCarouselLiteOptions['axis'], position: number): void {
    const translate = axisToTranslateMap[axis];

    element.style.transform = translate(position);
}

const scrollOptions: ScrollToOptions = { behavior: 'auto', left: 0, top: 0 };

/**
 * Делает расчет следующей позиции карусели исходя из параметров
 * index, align и размеров элементов track и carousel.
 * После применяет transform к элементу track или вызывает scrollTo на элементе carousel.
 * При scrollMode равном translate Анимирование происходит из-за CSS свойства `transition-property: transform`,
 * применённому к элементу track
 *
 * @param index индекс элемента к которому нужно сделать transform
 * @param prevIndex индекс предыдущего активного элемента для расчёта дистанции между индексами
 * @param axis ось вдоль которой будет происходить transform
 * @param align определяет позицию активного элемента относительно элемента carousel
 * @param track элемент к которому применяется transform
 * @param carousel элемент содержащий track
 * @param disableAnimation флаг для отключения анимирования прокрутки
 * @param scrollMode прокрутка через scrollTo или через translate
 */
export function translateToIndex(
    index: number,
    prevIndex: number,
    axis: UseCarouselLiteOptions['axis'],
    align: NonNullable<UseCarouselLiteOptions['scrollAlign']>,
    track: HTMLElement | null,
    carousel: HTMLElement | null,
    disableAnimation: boolean,
    scrollMode: NonNullable<UseCarouselLiteOptions['scrollMode']>,
): void {
    if (track === null || carousel === null) {
        return;
    }

    const scrollKey = axisToScrollKeyMap[axis];

    if (disableAnimation === false && scrollMode === 'translate' && carousel[scrollKey] !== 0) {
        carousel.scrollTo(scrollOptions);
    }

    const itemToTranslateTo = track.children.item(index) as HTMLElement | null;

    if (itemToTranslateTo === null) {
        return;
    }

    const numberOfItems = track.children.length;

    const offsetKey = axisToOffsetKeyMap[axis];
    const sizeKey = axisToSizeKeyMap[axis];

    const carouselSize = carousel[sizeKey];
    const trackSize = track[sizeKey];
    const trackOffset = track[offsetKey];
    const itemSize = itemToTranslateTo[sizeKey];
    const itemStart = itemToTranslateTo[offsetKey];

    const translatePosition = getTranslatePosition(itemSize, itemStart, carouselSize, trackOffset, align);

    const position = boundPosition(translatePosition, trackSize, carouselSize, trackOffset);

    if (scrollMode === 'scroll') {
        carousel.scrollTo({
            behavior: disableAnimation ? 'auto' : 'smooth',
            left: axis === 'x' ? position : 0,
            top: axis === 'y' ? position : 0,
        });

        return;
    }

    if (disableAnimation === true || isMaximumDistance(prevIndex, index, numberOfItems) === true) {
        // Выключаем стандартный easing, переключая его на step-start
        setTimingFunction(track, 'step-start');

        translateToPosition(track, axis, position);

        /**
         * Включаем стандартный easing.
         * нужно делать через setTimeout, чтобы transform успел произойти до изменения transitionTimingFunction
         */
        setTimeout(setTimingFunction, 0, track, '');

        return;
    }

    translateToPosition(track, axis, position);
}
