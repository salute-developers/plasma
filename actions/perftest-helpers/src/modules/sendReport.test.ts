/* eslint-disable no-multi-assign, @typescript-eslint/no-explicit-any */
import { jest } from '@jest/globals';
import * as got from 'got';

import * as fsUtils from '../utils/fs';

import PerftestSendReportApi from './sendReport';

jest.mock('crypto', () => ({
    randomUUID: () => 'asd',
}));

describe('perftest/sendReport', () => {
    afterEach(() => {
        jest.clearAllMocks();
        jest.restoreAllMocks();
        jest.resetModules();
    });

    describe('send', () => {
        let readJsonMock: jest.SpiedClass<any>;
        let gotMock: jest.SpiedClass<any>;
        let transformMock: jest.Mock;

        class ApiWithoutTransform extends PerftestSendReportApi {
            protected transform = (transformMock = jest.fn() as any);
        }

        beforeEach(() => {
            gotMock = jest.spyOn(got, 'default' as any);
            readJsonMock = jest.spyOn(fsUtils, 'readJson') as any;
        });

        it('should not send if version changed', async () => {
            readJsonMock.mockResolvedValue({ isVersionChanged: true });

            await new ApiWithoutTransform().send({ reportPath: 'azaza' } as any);

            expect(transformMock).not.toHaveBeenCalled();
            expect(gotMock).not.toHaveBeenCalled();
        });

        it('should not send if result is empty', async () => {
            readJsonMock.mockResolvedValue({ isVersionChanged: false, result: {} });

            await new ApiWithoutTransform().send({ reportPath: 'azaza' } as any);

            expect(transformMock).not.toHaveBeenCalled();
            expect(gotMock).not.toHaveBeenCalled();
        });

        it('should call this.transform if input is ok', async () => {
            const jsonReport = { isVersionChanged: false, result: { foo: 'bar' }, staticTaskChange: { baz: 'kl' } };
            const sendParams = {
                reportPath: 'azaza',
                subject: 'kek',
                commitHash: '12354',
                referrer: 'http://example.com',
            };
            const body = [{ top: 'kek' }];
            const api = new ApiWithoutTransform();

            readJsonMock.mockResolvedValue(jsonReport);
            transformMock.mockImplementation(() => body);
            gotMock.mockResolvedValue({ statusCode: 200 });

            await api.send(sendParams);

            expect(transformMock).toHaveBeenCalledWith({
                result: jsonReport.result,
                staticTaskChange: jsonReport.staticTaskChange,
                subject: sendParams.subject,
                commitHash: sendParams.commitHash,
                referrer: sendParams.referrer,
            });

            expect(gotMock).toHaveBeenCalledTimes(1);
            expect(gotMock.mock.calls[0][0]).toEqual(api.METRICS_URL);
            expect((gotMock.mock.calls[0][1] as any).method).toEqual('POST');
            expect((gotMock.mock.calls[0][1] as any).body).toEqual(JSON.stringify(body));
        });
    });

    describe('transform', () => {
        let getPayloadMock: jest.Mock;

        class ApiWithoutGetPayload extends PerftestSendReportApi {
            protected getPayload = (getPayloadMock = jest.fn() as any);

            public transform = super.transform;
        }

        it('should receive parameters and transform them the right way', () => {
            const api = new ApiWithoutGetPayload();
            const result = {
                kek: {
                    render: {
                        mean: {
                            old: [1, 1] as [number, number],
                            new: [2, 2] as [number, number],
                            change: {
                                percentage: 100,
                                difference: 1,
                                significanceRank: 'low' as const,
                            },
                        },
                    },
                },
            };
            getPayloadMock.mockReturnValue(1);

            const transformed = api.transform({
                result,
                staticTaskChange: result.kek,
                subject: '123',
                commitHash: 'qwrqw',
                referrer: 'http://example.com',
            });
            const firstRunId = transformed[0].runId;

            expect(typeof firstRunId).toEqual('string');
            expect(transformed.every(({ runId }) => runId === firstRunId)).toBe(true);

            transformed.forEach((record) => delete (record as any).runId);

            expect(transformed).toMatchSnapshot();
        });
    });

    describe('getPayload', () => {
        class ApiWithExposedGetPayload extends PerftestSendReportApi {
            public getPayload = super.getPayload;
        }

        it('should accept number', () => {
            const result = new ApiWithExposedGetPayload().getPayload(123);

            expect(result).toMatchSnapshot();
        });

        it('should accept pair of numbers', () => {
            const result = new ApiWithExposedGetPayload().getPayload([420, 123]);

            expect(result).toMatchSnapshot();
        });

        it('should accept dict', () => {
            const result = new ApiWithExposedGetPayload().getPayload({
                top: 'kek',
                wav: 123,
            });

            expect(result).toMatchSnapshot();
        });
    });
});
