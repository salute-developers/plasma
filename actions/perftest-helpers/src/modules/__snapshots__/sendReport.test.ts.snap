// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`perftest/sendReport getPayload should accept dict 1`] = `"{"top":"kek","wav":"123"}"`;

exports[`perftest/sendReport getPayload should accept number 1`] = `"{"0":"123"}"`;

exports[`perftest/sendReport getPayload should accept pair of numbers 1`] = `"{"0":"420","1":"123"}"`;

exports[`perftest/sendReport transform should receive parameters and transform them the right way 1`] = `
[
  {
    "commitHash": "qwrqw",
    "componentId": "kek",
    "kind": "old",
    "metricId": "mean",
    "payload": 1,
    "referrer": "http://example.com",
    "service": "123",
    "taskId": "render",
  },
  {
    "commitHash": "qwrqw",
    "componentId": "kek",
    "kind": "new",
    "metricId": "mean",
    "payload": 1,
    "referrer": "http://example.com",
    "service": "123",
    "taskId": "render",
  },
  {
    "commitHash": "qwrqw",
    "componentId": "kek",
    "kind": "change",
    "metricId": "mean",
    "payload": 1,
    "referrer": "http://example.com",
    "service": "123",
    "taskId": "render",
  },
  {
    "commitHash": "qwrqw",
    "componentId": "staticTaskChange",
    "kind": "old",
    "metricId": "mean",
    "payload": 1,
    "referrer": "http://example.com",
    "service": "123",
    "taskId": "render",
  },
  {
    "commitHash": "qwrqw",
    "componentId": "staticTaskChange",
    "kind": "new",
    "metricId": "mean",
    "payload": 1,
    "referrer": "http://example.com",
    "service": "123",
    "taskId": "render",
  },
  {
    "commitHash": "qwrqw",
    "componentId": "staticTaskChange",
    "kind": "change",
    "metricId": "mean",
    "payload": 1,
    "referrer": "http://example.com",
    "service": "123",
    "taskId": "render",
  },
]
`;
